### PR TITLE
Support the on-demand loading of blobs during iteration

### PR DESCRIPTION
**Pull request by ltamasi**
Summary: Currently, when using range scans with BlobDB, the iterator logic eagerly loads values from blob files when landing on a new entry. This can be wasteful in use cases where the values associated with some keys in the range are not used by the application. The patch introduces a new read option `allow_unprepared_value`; when specified, this option results in the above eager loading getting bypassed. Values needed by the application can be then loaded on an on-demand basis by calling the new iterator API `PrepareValue`. Note that currently, only regular single-CF iterators are supported; multi-CF iterators and transactions will be extended in later PRs.

Differential Revision: D64360723

### CODE REVIEW

File: db/db_iter.cc [hunk 1]
```diff
@@ -289,6 +305,24 @@ bool DBIter::SetValueAndColumnsFromMergeResult(const Status& merge_status,
   return true;
 }
 
+bool DBIter::PrepareValue() {
+  assert(valid_);
+
+  if (lazy_blob_index_.empty()) {
+    return true;
```
Comment (jowlyzhang) [inline db/db_iter.cc:312]: This is in case `PrepareValue` is called repeatedly, right? Shall we return `valid_` here?

**Review by jowlyzhang — COMMENTED** [review]
Thanks for the change, I only have some minor questions.

File: include/rocksdb/iterator_base.h [hunk 2]
```diff
@@ -74,6 +74,21 @@ class IteratorBase : public Cleanable {
...
+  // an on-demand basis.
+  //
+  // Returns true on success. Returns false and sets Valid() to false and
+  // status() to non-OK if there is an error while loading the value.
+  //
+  // Note: this method is currently only applicable to large values stored in
+  // blob files using BlobDB, and has no effect otherwise.
+  //
+  // REQUIRES: Valid()
+  virtual bool PrepareValue() { return true; }
```
[the remark this answers is not in the document]
Comment (ltamasi) [reply include/rocksdb/iterator_base.h:90]: Good point! Yes, the idea is that if `allow_unprepared_value` is true, applications are expected to call `PrepareValue` before using the value for any entry. (`PrepareValue` might be a no-op though e.g. if it's not a blob.) Let me update the comment, thanks!

File: db/db_iter.cc (hunk 1 above)
Comment (ltamasi) [reply db/db_iter.cc:312]: It also covers the case when the value is not a blob (e.g. BlobDB is actually disabled, or the value size is below `min_blob_size`).

Re: returning `valid_`, `PrepareValue` should only be called if the iterator is valid; this is documented in the API comment and we also have an assertion above.

**Review by jowlyzhang — APPROVED** [review]
Nice feature! Thank you for adding this improvement.

**Comment by ltamasi** [comment]
Thanks a lot for the review!

**Pull request closed without merging**